### ISSUE TITLE
Sending excessive STREAM_DATA_BLOCKED frames on long data transfer with curl

### ISSUE THREAD

**Issue by ygina** [cloudflare/quiche#1384]
**tl;dr** I am using `quiche` with `curl` to POST a long data transfer (1MB) in an experimental setting. I noticed poor performance when limiting packet queue size based on the _number_ of packets, and was looking into why. I think it's because `curl` is sending many excessive STREAM_DATA_BLOCKED frames, bloating the queue.

**Problem:** In Wireshark, I observed that the client initially sent some ~MTU-sized packets of ~1200 bytes, followed by _many_ ~45 byte packets. Logging the frames sent in one run showed that of 18164 total frames, 7.8% were STREAM frames, 91.6% STREAM_DATA_BLOCKED frames, and the remainder were ACK, CRYPTO, and PADDING frames. The STREAM_DATA_BLOCKED frames were sent in their own packets, aka not embedded in a STREAM packet like ACK frames were, and were bloating the queue.

I am not super familiar with the details of the [QUIC RFC](https://datatracker.ietf.org/doc/html/draft-ietf-quic-transport-23), but 19.13 only says "the sender **should** send STREAM_DATA_BLOCKED when it wishes to send data but is unable to due to stream-level flow control." The initial send and receive buffer size is `(1 * 1024 * 1024)` which is 1MB, so that seems fine. I find that the stream capacity is typically limited not by this initial buffer size, but by this variable `cwin_available` that is based on the cwnd set in quiche's recovery code.

**Proposal:** I don't see what the peer can do with this frame in a long data transfer within these reasonable parameters, so I made this fix on a fork (https://github.com/ygina/quiche/commit/fcc4c3f9a2fe0e149dd24889c4d4fffde2d01eaa) that just comments out the check in `send_body()`. And the RFC only says **should** and not **must**. This fix reduced the median time over 10 trials on a 152ms RTT link from 2.6s to 2.1s, and reduced the number of frames from 18164 to 1533.

**Questions:** I'd like to check if my understanding of the STREAM_DATA_BLOCKED frame is correct. Would this fix be useful in other settings? Is this an issue with the curl code calling `send_body()` with too large a body and not quiche itself? Can we still send STREAM_DATA_BLOCKED frames but fewer when the stream's capacity is limited, such as by piggybacking this frame on STREAM packets?

**Comment by LPardue** [comment]
Hi @ygina, thanks for the report. 

STREAM_DATA_BLOCKED means that the stream window credits available to a sender are not sufficient to send the amount of data that they would have liked. The frame can potentially be used as a signal to indicate that thr transport layer is throttling an application's send rate, the peer could then choose to open up an even larger window than it might otherwise.

If a stream is detected blocked, there is the possibility that we can't send any STREAM frames. That would limit us to coalescing STREAM_DATA_BLOCKED frames in packets with other frames. But if there are no other frames to send, we'd end up with these packets containing just blocked signals. That's not wrong per the protocol, but it does seem that it could be optimised.

Furthermore, it does seem that its possible for us to send blocked frames excessively in some scenarios and it would be nice to reduce that.

The cwnd check is made to avoid us buffering data into quiche that can't be sent on the wire due to congestion control. This is backpressure and can be useful for applications that are generating dynamic content or streaming it from another source. So I wouldn't want to remove that check.

I think the initial approach to addressing this would be for us to emit a STREAM_DATA_BLOCKED only when the flow control window is fully utilised, and only submit subsequent STREAM_DATA_BLOCKED frames after the window has increased.

**Comment by LPardue** [comment]
@ygina could you please try out https://github.com/cloudflare/quiche/pull/1386 and see if it addresses the issues you observed?

**Comment by ygina** [comment]
Hi @LPardue thanks for the quick response. In the below table, A is the original code, B is with the check commented out as in my fork, and C is with the commit from the `stream-blocked-only-when-flow-control-blocked branch` instead of commenting out the check. It does not seem to have improved the situation significantly.

```
    1MB upload (s)     # frames sent
A   2.108151           1467
B   2.605958           10551
D   2.610132           9508
```

**Comment by LPardue** [comment]
Ah yeah, I see what's happening now, this is caused by the behaviour in the H3 layer.

I'm not super happy with the idea of disabling STREAM_DATA_BLOCKED altogether, a SHOULD requirement is a pretty strong mandate that this is how QUIC is intended to work, and never sending the frame might have unanticipated consequences. Something that we could consider doing is to support a configurable rate of sending, i.e. if the blocked signal needs to be sent, don't send more than X STREAM_DATA_BLOCKED-frame-containing-packet per Y total sent packets. Apps that really wanted to, could make it so that the frames are never sent.

**Comment by ygina** [comment]
> The cwnd check is made to avoid us buffering data into quiche that can't be sent on the wire due to congestion control. 

I see, the check I removed marks streams as blocked which subsequently causes STREAM_DATA_BLOCKED frames to be sent.

> Ah yeah, I see what's happening now, this is caused by the behaviour in the H3 layer.

Could you elaborate on this?

> Something that we could consider doing is to support a configurable rate of sending, i.e. if the blocked signal needs to be sent, don't send more than X STREAM_DATA_BLOCKED-frame-containing-packet per Y total sent packets. Apps that really wanted to, could make it so that the frames are never sent.

This seems like it would address my issue! How would an application decide how to set this rate? In my contrived setting, I know that the peer would never open up a larger window and so it's pointless to send STREAM_DATA_BLOCKED frames. But in a real setting, how frequently makes sense? Probably somewhere between never and 10x for every successfully sent packet...

**Comment by junhochoi** [comment]
> don't send more than X STREAM_DATA_BLOCKED-frame-containing-packet per Y total sent packets. Apps that really wanted to, could make it so that the frames are never sent.

I think we can try to send `STREAM_DATA_BLOCKED` only when `SendBuf.blocked_at` moved.

**Comment by LPardue** [comment]
I made an alternative PR that explains the problem in detail; see https://github.com/cloudflare/quiche/pull/1391/files

The TL;DR is that the h3 layer was using stream_writable in a way that could skirt the dupe STREAM_DATA_BLOCKED checks. This was due to an oversight and the checks should be done in both places. 

With the fix, my tests show the rate of STREAM_DATA_BLOCKED frames drop by two orders of magnitude. That seems good enough for me. If we want to go further and make a special API, that would be a feature request that needs some more consideration.

@ygina could you please test again with https://github.com/cloudflare/quiche/pull/1391/files and see if you also see a reduction in frame count?

**Comment by ygina** [comment]
Yea! Time (~2.1s) and number of packets (~1400) are now comparable with this fix.
